### PR TITLE
Annotations for .data_vars() and .coords()

### PR DESCRIPTION
**Pull request by crusaderky**

### CODE REVIEW

File: xarray/core/coordinates.py [hunk 1]
```diff
@@ -23,49 +34,58 @@
...
+    def __setitem__(self, key: Hashable, value: Any) -> None:
         self.update({key: value})
 
     @property
-    def indexes(self):
+    def _names(self) -> Set[Hashable]:
+        raise NotImplementedError()
+
+    @property
+    def dims(self) -> Union[Mapping[Hashable, int], Tuple[Hashable, ...]]:
```
Comment (max-sixty) [inline xarray/core/coordinates.py]: Why the change here from `return self._data.dims`?

File: xarray/core/coordinates.py [hunk 2]
```diff
@@ -23,49 +34,58 @@
...
-    def __setitem__(self, key, value):
+    def __getitem__(self, key: Hashable) -> "DataArray":
+        raise NotImplementedError()
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
         self.update({key: value})
 
     @property
-    def indexes(self):
+    def _names(self) -> Set[Hashable]:
```
Comment (max-sixty) [inline xarray/core/coordinates.py]: Is this new? Should this be on the abstract class?

File: xarray/core/coordinates.py [hunk 3]
```diff
@@ -259,19 +291,15 @@ def _update_coords(self, coords):
     def variables(self):
         return Frozen(self._data._coords)
 
-    def _to_dataset(self, shallow_copy=True):
```
Comment (max-sixty) [inline xarray/core/coordinates.py]: I think removing this should be fine...

File: xarray/core/coordinates.py (hunk 2 above)
Comment (crusaderky) [reply xarray/core/coordinates.py]: It's a property invoked by the methods of the abstract class, defined in both the implementations.

File: xarray/core/coordinates.py (hunk 1 above)
Comment (crusaderky) [reply xarray/core/coordinates.py]: Because ``DataArray.dims`` is a ``Tuple[Hashable, ...]`` whereas ``Dataset.dims`` is a ``Mapping[Hashable, int]``

Comment (crusaderky) [reply xarray/core/coordinates.py]: See below for the two implementations, which are identical in code but different in signature.
The method needs stay in the abstract class because other methods of the abstract class invoke it; this signature with NotImplementedError() guarantees that the methods of the abstract class can only use dims as the common denominator of the two output types, which is a ``Collection[Hashable]``.

File: xarray/core/coordinates.py (hunk 3 above)
Comment (crusaderky) [reply xarray/core/coordinates.py]: I searched and there's nothing invoking it across the whole codebase

Comment (max-sixty) [reply xarray/core/coordinates.py]: Nice

File: xarray/core/coordinates.py (hunk 1 above)
Comment (max-sixty) [reply xarray/core/coordinates.py]: OK great

File: xarray/core/coordinates.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/coordinates.py]: Here, I'm not so sure - do all concrete classes need to define this? Generally we wouldn't define private methods on an abstract class (even if all currently concrete classes implement it), because the implementation would be up to the concrete class?

Comment (crusaderky) [reply xarray/core/coordinates.py]: What I basically did was turn AbstractCoordinates from a mixin (that is, a class that invokes own methods that are only defined in a subclass) into a C++-style abstract class.
mypy, justifiably, doesn't like mixins - short of hacking your way through with ``# type`` tags.

This, in synthesis, is how it was before:
```
class C:
    def g(self) -> int:
        return self.f()


class D(C):
    def f(self) -> int:
        return 123
```
mypy correctly states:
```3: error: "C" has no attribute "f"```

Comment (max-sixty) [reply xarray/core/coordinates.py]: Right - I agree this should either be an abstract class without implementations or a mixin with implementations - not a bit of each. It's a bit odd we call this class `Abstract` but then enforce a `_names` property.

That a nice upside of mypy - we get to find those things out

I agree we're in a reasonable local maximum here.

Comment (shoyer) [reply xarray/core/coordinates.py]: Yes, thanks for fixing this up!
